keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are many issues with today's Eviction API and PodDisruptionBudgets that are of great concern
+to cluster admins and application owners. These issue range from insufficient data safety or
+application availability to node draining issues.
+
+This KEP proposes to add a new declarative Evacuation API to manage the evacuation of pods. Its
+mission is to allow for a cooperative evacuation (removal) of a pod, usually in order to run the
+pod on another node. If the owner of the pod does not cooperate, the evacuation will try to resort
+to a pod eviction (API initiated Eviction).
+
+This new API can be used to implement additional capabilities around node draining, pod
+descheduling, or as a general interface between applications and/or controllers. It would provide
+additional safety and observability guarantees and prevent bad practices as opposed to just using
+the current Eviction API endpoint and PodDisruptionBudgets.
+
+## Motivation
+
+Many of today's solutions rely on eviction (API-initiated Eviction) as the goto-safe way to remove
+a pod from a node (kubectl drain, descheduler, cluster autoscaler, partially scheduler preemption).
+Unfortunately, this is done in an application agnostic way and can cause many problems.
+
+From an application owner or developer perspective, the only standard tool they have to protect
+them against eviction is a PodDisruptionBudget. This is sufficient in a basic scenario with a simple
+multi-replica application. The edge case applications, where this does not work are very important
+to the cluster admin or controllers managing workload distribution on nodes, as they can for
+example block the node drain. And, in turn, very important to the application owner, as the admin
+can then override the pod disruption budget and disrupt their sensitive application anyway.
+
+The major issues are:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
+   experiencing application downtime during the node drain. They cannot use PDBs with
+   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
+   needs high availability either, due to a preference for a simpler deployment model, lack of
+   application support for HA, or to minimize compute costs. Also, any automated solution needs
+   to edit the PDB to account for the additional pod that needs to be spun to move the workload
+   from one node to another. This has been discussed in issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
```

Comment:
As far as I have read about EvacuationAPI, it seems to respect PDBs always.
But I wish to know, will EvacuationAPI be able to disable PDBs or ignore PDBs to help the case where the operator expects zero living nodes in the cluster (i.e. shutting down cluster during the weekends/holidays for maintenance and starting them again after the weekend)? 
The above case requires all nodes to shut down,  and having PDBs with `minAvailable: 1` or `maxUnavailable: 0` will definitely block the draining the nodes.

Reply:
the feature is not limited by PDBs, the workloads can use them but they can also bypass them if there is a controller that can make more advanced decisions

Reply:
But for the k8s built-in controllers like deployments, statefulsets, etc; if a PDB was configured, how will EvacuationAPI be able to bypass the disruption budget and allow node termination below the PDB configured values?

Reply:
I will update the KEP soon. But the mechanism of optionally bypassing the PDB should not change, and requires action from the evacuator.. PTAL at the KEP or just the attached diagram: https://github.com/kubernetes/enhancements/pull/4565/files#diff-560381a7190e130fded9e8758d53d2e5c4b69c39838485fc7a3fb13be096ae47

Reply:
looking at the diagram, where in the innermost loop, it says the Evacuator updates the Evacuation status. I just want to clarify the meaning of the part where it says 

> "make sure the Evacuation status is up to date, evict the pod **if it is not**". 

Does it mean if the evacuator failed to update the Evacuation status for a specific time, the pod will be terminated regardless?

Reply:
yup, this is implemented via the `.spec.progressDeadlineSeconds`, but the min/default values should be forgiving to allow for evacuator disruptions

Reply:
oh thank you. As I see it, does `.spec.progressDeadlineSeconds` double as the polling duration counting from the most recent evacuator update, waiting for other potential instigators?

Reply:
No, the `.spec.progressDeadlineSeconds` is a static value that cannot be changed. Every interceptor/evacuator abides by the same deadline to produce an update tick. By making the value immutable, we can make life easier for the interceptors to not miss the deadline.